### PR TITLE
Fix exchange issue, added tests. 

### PR DESCRIPTION
**Pull request by willjgriff**
Add functionality for converting tokens to ETH when less tokens are available than are required.

Added tests.

### CODE REVIEW

File: test/brightIdFaucet.js [hunk 1]
```diff
@@ -1,29 +1,499 @@
...
+        const { addressVoid: newAddressVoid } = await brightIdFaucet.claimers(faucetUserSecondAddress)
+        assert.isTrue(originalAddressVoid, 'Incorrect original address void')
+        assert.isFalse(newAddressVoid, 'Incorrect new address void')
+      })
+
+      it('voids all previously registered accounts when already voided accounts', async () => {
+        addresses = [faucetUserSecondAddress, faucetUser]
+        sig = getVerificationsSignature(addresses, timestamp)
+        await brightIdFaucet.claimAndOrRegister(BRIGHT_ID_CONTEXT, addresses, timestamp, sig.v, sig.r, sig.s, { from: faucetUserSecondAddress })
+        addresses = [faucetUserThirdAddress, faucetUserSecondAddress, faucetUser]
```
Comment (fabriziovigevani) [inline test/brightIdFaucet.js:234]:
```suggestion
        addresses = [faucetUserThirdAddress, ...addresses]
```

File: test/brightIdFaucet.js [hunk 2]
```diff
@@ -1,29 +1,499 @@
...
+          assert.isTrue(token2BalanceAfter.gt(token2BalanceBefore), 'Incorrect token balance, too low')
+          await web3.eth.sendTransaction({ from: faucetUserSecondAddress, to: faucetUser, value: eth1BalanceToRemove })
+          await web3.eth.sendTransaction({
+            from: faucetUserSecondAddress,
+            to: secondFaucetUser,
+            value: eth2BalanceToRemove
+          })
+        })
+
+        it('converts faucet token to eth when exchange will cost more than faucet payment', async () => {
```
Comment (fabriziovigevani) [inline test/brightIdFaucet.js:393]: Might be a bit clearer

```suggestion
        it('converts full faucet token payment to eth when exchange will cost more than faucet payment', async () => {
```

File: contracts/BrightIdFaucet.sol [hunk 3]
```diff
@@ -126,68 +163,65 @@ contract BrightIdFaucet is Ownable {
...
+        require(msg.sender == _addrs[0], ERROR_SENDER_NOT_IN_VERIFICATION);
+        require(!claimer.addressVoid, ERROR_ADDRESS_VOIDED);
 
-        claim();
+        uint256 currentPeriod = getCurrentPeriod();
+        if (_canClaim(claimer, currentPeriod)) {
+            _claim(claimer, currentPeriod);
+        }
 
         uint256 nextPeriod = getCurrentPeriod() + 1;
```
[the remark this answers is not in the document]
Comment (willjgriff) [reply contracts/BrightIdFaucet.sol:177]: I'll leave this as done in the next PR.

File: contracts/BrightIdFaucet.sol [hunk 4]
```diff
@@ -220,13 +254,38 @@ contract BrightIdFaucet is Ownable {
     }
 
     function _canClaim(Claimer storage claimer, uint256 currentPeriod) internal view returns (bool) {
-        bool userRegisteredCurrentPeriod = currentPeriod > 0 && claimer.registeredForPeriod == currentPeriod;
```
[the remark this answers is not in the document]
Comment (willjgriff) [reply contracts/BrightIdFaucet.sol:223]: Yeah, `claimer.latestClaimPeriod` can not be less than `currentPeriod` when `currentPeriod` is 0.

File: test/brightIdFaucet.js [hunk 5]
```diff
@@ -1,29 +1,499 @@
...
+  let token, tokenExchange
 
-    console.log(await brightIdFaucet.claimers(faucetOwner))
+  beforeEach(async () => {
+    token = await Token.new('Test Token', 'TTN')
   })
 
+  const createUniswapExchangeForToken = async (token, tokenLiquidity, ethLiquidity) => {
+    const tokenLiquidityBn = toBnWithDecimals(tokenLiquidity)
+    const ethLiquidityBn = toBnWithDecimals(ethLiquidity)
```
[the remark this answers is not in the document]
Comment (willjgriff) [reply test/brightIdFaucet.js:35]: Done in next PR.

**Pull request merged**